### PR TITLE
Name regex doesn't match hyphen, dots and quote #41

### LINKED ISSUE #41: Name regex doesn't match hyphen, dots and quote
For that matter, name regx doesn't match any non-ascii names

### PR DESCRIPTION
**Pull request by okkhoy**
Fixes #41

### CODE REVIEW

**Comment by damithc** [comment]
@ndt93 for your testing and merging please.

File: test/input.txt [hunk 1]
```diff
@@ -147,6 +147,19 @@
   # clears all
   clear
   list
+  
+##########################################################
+# test new regex for name
+##########################################################	 
+  add Name with'arc p/99999999 e/name@arc.com a/1 long, #01-valid address, 890765
```
Comment (ndt93) [inline test/input.txt]: You may want to add more test commands or modify this test here to cover additional details described by the new regex i.e. `.`, `-` and unicode letters.

**Comment by ndt93** [comment]
@okkhoy I made a comment on the test case. Also, can you help update your branch to the current master? Thanks.

**Comment by damithc** [comment]
@okkhoy no hurry. You can resume this when you have time.

**Comment by okkhoy** [comment]
@ndt93 please check the new tests and merge @damithc

**Comment by ndt93** [comment]
@okkhoy Do the tests pass on your side? It's failing on my side (macOS)

**Comment by okkhoy** [comment]
yes. they do. may be due to line ending? works fine on windows.

**Comment by ndt93** [comment]
The outputs are different between actual and expected for me. Here is the diff:
<pre>
327,328c327,328
< || Enter command: || [Command entered:  add Bjï¿½rn Borg p/98765432 e/borg@gmail.com a/311, Clementi Ave 2, #02-25]
< || Person names should be spaces or alphabetic characters
---
> || Enter command: || [Command entered:  add Björn Borg p/98765432 e/borg@gmail.com a/311, Clementi Ave 2, #02-25]
> || New person added: Björn Borg Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
333a334
> || 	4. Björn Borg Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
335c336
< || 3 persons listed!
---
> || 4 persons listed!
337,338c338,339
< || Enter command: || [Command entered:  add Josï¿½ Eduardo p/98765432 e/borg@gmail.com a/311, Clementi Ave 2, #02-25]
< || Person names should be spaces or alphabetic characters
---
> || Enter command: || [Command entered:  add José Eduardo p/98765432 e/borg@gmail.com a/311, Clementi Ave 2, #02-25]
> || New person added: José Eduardo Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
343a345,346
> || 	4. Björn Borg Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
> || 	5. José Eduardo Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
345c348
< || 3 persons listed!
---
> || 5 persons listed!
</pre>

**Comment by ndt93** [comment]
Seems like a unicode problem...

File: test/expected.txt [hunk 2]
```diff
@@ -297,6 +297,78 @@
...
+|| Enter command: || [Command entered:  list]
+|| 	1. Name with'arc Phone: 99999999 Email: name@arc.com Address: 1 long, #01-valid address, 890765 Tags: 
+|| 	2. John-Doe Phone: 98765432 Email: johnd@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
+|| 	3. John-Doe jr. Phone: 98765432 Email: johnd@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
+|| 
+|| 3 persons listed!
+|| ===================================================
+|| Enter command: || [Command entered:  add Bj�rn Borg p/98765432 e/borg@gmail.com a/311, Clementi Ave 2, #02-25]
+|| New person added: Bj�rn Borg Phone: 98765432 Email: borg@gmail.com Address: 311, Clementi Ave 2, #02-25 Tags: 
+|| ===================================================
```
Comment (ndt93) [inline test/expected.txt]: @okkhoy Some characters in the name are corrupted so I think the encoding is not in UTF-8. Can you help check this again?

Comment (ndt93) [inline test/expected.txt]: There is the same encoding problem in `input.txt`.

**Comment by okkhoy** [comment]
Any suggestions to fix this? works fine on my linux machine as well :-/

**Comment by ndt93** [comment]
@okkhoy The characters encoding for the files is currently `iso-8859-1`. Is there any option in the your text editors or tools (that you use to save the output) on your machine that let us select the `UTF-8` encoding?

**Comment by ndt93** [comment]
@okkhoy Your files now contain the Byte Order Mark (BOM) which our application incorrectly reads as a command input.

image
image

@okkhoy @damithc This is a documented Java bug that it doesn't handle BOM correctly but is not fixed due to backward compatibility. Would you like try saving the files without BOM (since it's optional) or handle the BOM in the code?

**Comment by okkhoy** [comment]
oh! that is weird. I explicitly disabled BOM while converting the encoding
:-/

**Comment by damithc** [comment]
Would this cause problems in future when students edit the file? You know, 'The files looks same but tests are failing' kind of problems.

**Comment by ndt93** [comment]
If the students don't touch the `actual.txt` file then this shouldn't happen. If they edit or resave it then yup that might happen.

**Comment by damithc** [comment]
They might have to change that file. Is this problem specific to this PR or is it a problem that existed from the beginning?

**Comment by ndt93** [comment]
This problem exists from the beginning but only manifests in this PR because we now accept unicode characters and are trying to save the files in `UTF-8`. The BOM is optional and only exists for certain encodings.

**Comment by damithc** [comment]
In that case perhaps we should apply this enhancement to level 4 instead? It should not be a problem at level 4 right? Level 2 can survive without this feature.

**Comment by ndt93** [comment]
@okkhoy Ok then can you help save the files without the BOM? Thanks.

**Pull request merged**